lib/internal/quic/core.js

```diff
@@ -0,0 +1,3341 @@
+
+    if (this.listenerCount('pathValidation') > 0)
+      toggleListeners(handle, 'pathValidation', true);
+
+    if (this.listenerCount('usePreferredAddress') > 0)
+      toggleListeners(handle, 'usePreferredAddress', true);
+
+    this.#maybeReady(0x2);
+  }
+
+  [kSocketReady]() {
+    this.#maybeReady(0x1);
+  }
+
+  // The QuicClientSession is ready for use only after
+  // (a) The QuicSocket has been bound and
+  // (b) The internal handle has been created
+  #maybeReady = function(flag) {
+    this.#ready |= flag;
+    if (this.ready)
+      process.nextTick(emit.bind(this, 'ready'));
+  };
+
+  get allowEarlyData() {
+    return this.#allowEarlyData;
+  }
+
+  get ready() {
+    return this.#ready === 0x3;
+  }
+
+  get handshakeStarted() {
+    return this.#handshakeStarted;
+  }
+
+  startHandshake() {
+    if (this.destroyed)
+      throw new ERR_QUICSESSION_DESTROYED('startHandshake');
+    if (this.#handshakeStarted)
+      return;
+    this.#handshakeStarted = true;
+    if (!this.ready) {
+      this.once('ready', () => this[kHandle].startHandshake());
+    } else {
+      this[kHandle].startHandshake();
+    }
+  }
+
+  get ephemeralKeyInfo() {
+    return this[kHandle] !== undefined ?
+      this[kHandle].getEphemeralKeyInfo() :
+      {};
+  }
+
+  #setSocketAfterBind = function(socket, callback) {
+    if (socket.destroyed) {
+      callback(new ERR_QUICSOCKET_DESTROYED('setSocket'));
+      return;
+    }
+
+    if (!this[kHandle].setSocket(socket[kHandle])) {
+      callback(new ERR_QUICCLIENTSESSION_FAILED_SETSOCKET());
+      return;
+    }
+
+    if (this.socket) {
+      this.socket[kRemoveSession](this);
+      this[kSetSocket](undefined);
+    }
+    socket[kAddSession](this);
+    this[kSetSocket](socket);
+
+    callback();
+  };
+
+  setSocket(socket, callback) {
+    if (!(socket instanceof QuicSocket))
+      throw new ERR_INVALID_ARG_TYPE('socket', 'QuicSocket', socket);
+
+    if (typeof callback !== 'function')
+      throw new ERR_INVALID_CALLBACK();
+
+    socket[kMaybeBind](() => this.#setSocketAfterBind(socket, callback));
+  }
+}
+
+function streamOnResume() {
+  if (!this.destroyed)
+    this[kHandle].readStart();
+}
+
+function streamOnPause() {
+  if (!this.destroyed /* && !this.pending */)
+    this[kHandle].readStop();
+}
+
+class QuicStream extends Duplex {
+  #closed = false;
+  #aborted = false;
+  #defaultEncoding = undefined;
+  #didRead = false;
+  #id = undefined;
+  #highWaterMark = undefined;
+  #push_id = undefined;
+  #resetCode = undefined;
+  #session = undefined;
+  #dataRateHistogram = undefined;
+  #dataSizeHistogram = undefined;
+  #dataAckHistogram = undefined;
+  #stats = undefined;
+
+  constructor(options, session, push_id) {
+    const {
+      highWaterMark,
+      defaultEncoding,
+    } = options;
+    super({
+      highWaterMark,
+      defaultEncoding,
+      allowHalfOpen: true,
+      decodeStrings: true,
+      emitClose: true,
+      autoDestroy: false,
```

Comment:
You mentioned before that quic streams have a different lifecycle that does not exactly match the stream defaults (i.e. `autoDestroy`). Would you mind expanding a bit on that?

Reply:
A `QuicStream` cannot be destroyed until the underlying `ngtcp2_connection` explicitly signals that the stream is closed and will no longer be used. The default `autoDestroy` implementation causes the stream to be destroyed too early, causing an interruption in the protocol flow. So, we turn it off and wait for the explicit signal from ngtcp2 to destroy it and the underlying handle.

Or, put another way, simply closing the readable and writable sides of the Duplex does not mean we can destroy the resource.

Reply:
Why can't we wait for the signal in `_destroy`?

i.e.

```js
async _destroy(err, cb) {
  await ngtcp2_connection_close
  close handle
  cb()
}
```

Reply:
Or can the user perform any useful operation on the object between the stream completing and the ngtcp2_connection signal?

Reply:
Yes, there are operations that can be performed on the object. What is happening during the time between the sides of the duplex closing and the actual destroy is the underlying protocol ensuring reliable delivery of the data transmission. User code (or error handling on the QuicSession level) can short circuit that at any time by explicitly calling `destroy()` which forces the handle to be freed immediately and abandons the transmission ack attempt. There are also stats being actively collected on the object that are updated as data is being transmitted and acknowledged.

Reply:
Interesting. I need to give this more thought in the realm of streams but I think your implementation makes sense. Maybe a short version of your above description would be appropriate above `autoDestroy: false` to motivate why this option is used?